tests/unit/expand_conditions_spec.lua

```diff
@@ -0,0 +1,252 @@
+		describe("composite", function()
+			local function foo(b1, b2, b3)
+				-- Attention: use this only when testing (otherwise (pot.
+				-- malicious) users might inject code)
+				return helpers.exec_lua(
+					([[
+					local cond = require("luasnip.extras.expand_conditions").make_condition
+					local c = - ( cond(function() return %s end) + cond(function() return %s end) * cond(function() return %s end)) ^ ( cond(function() return %s end) + cond(function() return %s end) * cond(function() return %s end))
+					return c() == %s
+					]]):format(
+						tostring(b1),
+						tostring(b2),
+						tostring(b3),
+						tostring(b3),
+						tostring(b1),
+						tostring(b2),
+						tostring(not (b1 or b2 and b3) ~= (b3 or b1 and b2))
+					)
+				)
+			end
+			for _, ele in ipairs({
+				{ true, true, true },
+				{ true, true, false },
+				{ true, false, true },
+				{ true, false, false },
+				{ false, true, true },
+				{ false, true, false },
+				{ false, false, true },
+				{ false, false, false },
+			}) do
+				it(
+					("composite %s %s %s"):format(
+						tostring(ele[1]),
+						tostring(ele[2]),
+						tostring(ele[3])
+					),
+					function()
+						local test = function()
+							return foo(ele[1], ele[2], ele[3])
+						end
+						assert.has_no.errors(test)
+						assert.is_true(test())
+					end
+				)
+			end
+		end)
+	end)
+	describe("line_begin", function()
+		it("is at begin", function()
+			local function foo()
+				return helpers.exec_lua([[
+				local cond = require("luasnip.extras.expand_conditions")
+				local c = cond.line_begin
+				return not c("hello world", "hello world") ~= true -- allow nil/object
+				]])
+			end
+			assert.has_no.errors(foo)
+			assert.is_true(foo())
+		end)
+		it("is NOT at begin", function()
+			local function foo()
+				return helpers.exec_lua([[
+				local cond = require("luasnip.extras.expand_conditions")
+				local c = cond.line_begin
+				return not c("hello world", "ld") ~= false -- allow nil/object
+				]])
+			end
+			assert.has_no.errors(foo)
+			assert.is_true(foo())
+		end)
+	end)
+	describe("line_end", function()
+		it("is at begin", function()
+			local function foo()
+				return helpers.exec_lua([[
+				-- vim.api.nvim_get_current_line
+				vim = {api = {nvim_get_current_line = function() return "hello world ending" end}}
+				local cond = require("luasnip.extras.expand_conditions")
+				local c = cond.line_end
+				return not c("hello world ending") ~= true -- allow nil/object
+				]])
+			end
+			assert.has_no.errors(foo)
+			assert.is_true(foo())
+		end)
+		it("is NOT at begin", function()
+			local function foo()
+				return helpers.exec_lua([[
+				vim = {api = {nvim_get_current_line = function() return "hello world ending" end}}
+				local cond = require("luasnip.extras.expand_conditions")
+				local c = cond.line_end
+				return not c("hello world") ~= false -- allow nil/object
+				]])
+			end
+			assert.has_no.errors(foo)
+			assert.is_true(foo())
+		end)
+	end)
```

Comment:
I think these would be nicer as integration-tests, easier to maintain that way.

also:
faking `nvim_get_current_line` is a cool idea, the `vim`-table leaks to the following tests though (`extend_decorator` is missing `helpers.clear()` `before_each` :grimacing:)

Reply:
Well, I just tried to get around integration tests as it is tedious to always write these buffers for reference. Tbh I think it is easier to read this way where the relevant information (the content of the corresponding line) is right there.
Fixed `before_each` (hopefully, I'm still a bit unsure about the testing framework)

Reply:
Ohhhh that's my bad, you can use `screen:snapshot_util()` to capture the screen. Wow srry, I completely forgot about that, pita to write them by hand😅😬
Mhmmm proceed as you like, I'm fine with unit tests as well

Reply:
^^
Still I'd prefer the unit test tbh